add lch presentation option to color picker module

dterrahe opened the pull request:
This seems to get similar results to #6352 but in a more localised way.

The tooltip is getting rather large with this addition...

msdm commented:
I haven't tried this PR but if it gives similar results to mine (#6352), I think this one should be merged as it is simpler with fewer changes in the code. I was also planning to add HSL support and it would be nice if you could do that part as well.

dterrahe commented:
OK, I added HSL as well as a presentation option in the combobox and converted using dt_RGB_2_HSL. I have no idea if this is the right thing to do and if any scaling/clamping would be required.

However, I have two reservations here:
- I don't think it makes sense to add all possible transformations to the tooltip as well, since it would get huge. Maybe LCh shouldn't really be there either.
- Is this conceptually correct? i.e. does the min/max/mean after these transformation really correspond to a transformation of the rgb or lab statistics at the end, or should this be done at the pixel level? I have no opinion, just asking. If this is just for fun/comparability/indication of scale, then "accuracy" might not be that relevant. Doing at pixel level of course would require significantly more work (and possibly cause slowdown) so in that case the addition would need some justification.

msdm commented:
Thanks a lot

TurboGit commented:
>     * I don't think it makes sense to add all possible transformations to the tooltip as well, since it would get huge. Maybe LCh shouldn't really be there either.

Agreed, let's remove this from tooltip.

TurboGit commented on src/libs/colorpicker.c:
```diff
@@ -160,6 +161,18 @@ static void _update_sample_label(dt_colorpicker_sample_t *sample)
+    case 2:
+      // LCh
+      dt_Lab_2_LCH(lab, alt);
+      snprintf(text, sizeof(text), "%.02f %.02f %.02f", CLAMP(alt[0], .0f, 100.0f), alt[1], alt[2] * 360);
```
all format should be : %2.02f as LCh values can be > 10. otherwise the number does not align well.

msdm commented on src/libs/colorpicker.c:
```diff
@@ -164,8 +164,14 @@ static void _update_sample_label(dt_colorpicker_sample_t *sample)
+    case 3:
+      // HSL
+      dt_RGB_2_HSL(rgb, alt);
+      snprintf(text, sizeof(text), "%.02f %.02f %.02f", alt[0], alt[1], alt[2]);
```
The "H" channel (alt[0]) of HSL should be multiplied to 360 just like the "h" channel of LCh.

dterrahe commented:
Do we like a slightly bigger font for the RGB/Lab headers in the tooltip?

TurboGit commented:
All this looks good except the fact that the numbers are moving around because they are not using the same length in all color mode. Add three live sample and change the RVB, Lab, LCh... 

I would use everywhere: "%6.02f %6.02f %6.02f"

And probably also adjust the RGB format to be "%6d %6d %6d".

Last, the color for the RVB numbers makes the numbers hard to read. I would use standard white, better lisibility than "fun" effect to me. I understand that this has not been changed here, but maybe better to do that as were are at it.

TurboGit commented:
> Do we like a slightly bigger font for the RGB/Lab headers in the tooltip?

Looks ok to me.

dterrahe commented:
> I would use everywhere: "%6.02f %6.02f %6.02f"

Yes, definitely! Much calmer.

> And probably also adjust the RGB format to be "%6d %6d %6d".

Not sure; I see RGB as a double option, because it also makes the color patch larger, in case you don't care about the numbers.
If you really want to have all the numbers the same size, then I'd want to add another option "Patch", that simply doesn't print a number at all.

> Last, the color for the RVB numbers makes the numbers hard to read. I would use standard white, better lisibility than "fun" effect to me. I understand that this has not been changed here, but maybe better to do that as were are at it.

You are right (and I was wondering why you didn't bring that up before!) I have now changed these to brighter color number (#FF7F7F rather than just 'red'). I think this works quite well. Could go even brighter if needed. The only drawback now is that these colors this way are not themable, but they work well with all current themes imho.

I won't cry if you make me remove them after all, but I think it is fun and _almost_ useful this way.

TurboGit commented:
> and I was wondering why you didn't bring that up before!

I thought it would work ok, but after using it I found it not ok. So just changing my mind. I'll check the new colors on all themes and will come back.

TurboGit commented:
> then I'd want to add another option "Patch",

Not sure about this and I would still prefer to avoid jumping numbers and patch resized when going from RGB to Lab which are the two most commons options I'm using. But let's see what others have to say about that as it is really a matter of preference.

dterrahe commented:
Fair enough! You know I rely on you to catch my experiments that don't work. Means I don't have to self-censure (too much).

dterrahe commented:
> > then I'd want to add another option "Patch",
> 
> Not sure about this

I just tried it. Works quite well, especially with narrow panels when the text is unreadable anyway. And there's still the tooltip.

"None" is a better name. "Color Only" would be too long (for narrow panels). And there should be at least one character (to indicate where to aim for the tooltip). How about a hamburger? (☰) It might give the impression there is something clickable.

Bullseye? ◎ (U+25CE)

I like it. Can I have it? Please?

Completely separately, I had the idea of automatically adding a sample when clicking (or releasing after drawing an area) with CTRL held. It turns out this is somewhat problematic, because a continuous series of updates is sent to the function linked to darktable.lib->proxy.colorpicker.update_panel so many samples would be added. I know the + button can be linked to an accelerator, but that would permanently cost a key (and besides nobody does that before they decide they'd like it). This would need a bit of work, so I'm not sure if it is worth it.

dterrahe commented:
![image](https://user-images.githubusercontent.com/1549490/94065426-09150f00-fde3-11ea-8e99-f8c3a4a82c51.png)

johnny-bit commented:
bullseye looks great!

TurboGit commented:
> This would need a bit of work, so I'm not sure if it is worth it.

I would say no, to me this is not needed and would pollute the color picker lib when trying multiple time to select the proper area.

I'll merge this as-is. I'm not sure "none" is really useful (at least not to me) as the point of the color picker is to get the actual values of the pointed area. But maybe this will be useful for others and it does not make the GUI more complex. So let's this in.

One thing I've just seen is that there is the text on the tooltip "hover to highlight sample on canvas" but there is nothing happening. This is not a regression due to this PR, so can be fixed in another PR. I suspect a regression due to the pipe rewrite/optimization done recently.

TurboGit commented:
Thanks for this nice revamp of the color picker. Since 3.0 I think we have reworked almost every GUI part of dt.

dterrahe commented:
> I'll merge this as-is. I'm not sure "none" is really useful (at least not to me) as the point of the color picker is to get the actual values of the pointed area. But maybe this will be useful for others and it does not make the GUI more complex. So let's this in.

This would basically just be an easy access to the tooltip which has all the details (more than just the one statistic). The alternative would be to automatically switch to the bullseye when the text becomes unreadably small, but that would be more complicated...
